# *Pytorch로 단층 퍼셉트론 구현*

# 

### *데이터 만들기*

In [14]:
import torch
import torch.nn as nn

In [2]:
device = "cuda" if torch.cuda.is_available() else "cpu"
torch.manual_seed(777)
if device == "cuda":
    torch.cuda.manual_seed_all(777)

In [3]:
device

'cuda'

In [5]:
X = torch.Tensor([[0, 0], [0, 1], [1, 0], [1, 1]]).to(device)
Y = torch.Tensor([[0], [1], [1], [0]]).to(device)

In [6]:
print(X)
print(Y)

tensor([[0., 0.],
        [0., 1.],
        [1., 0.],
        [1., 1.]], device='cuda:0')
tensor([[0.],
        [1.],
        [1.],
        [0.]], device='cuda:0')


# 

### *1개의 뉴런을 가지는 시그모이드를 활용한 단층 퍼셉트론*

In [15]:
linear = nn.Linear(2, 1, bias=True)
print(f"linear : {linear}")

sigmoid = nn.Sigmoid()
print(f"sigmoid : {sigmoid}")

model = nn.Sequential(linear, sigmoid).to(device)
print(model)

linear : Linear(in_features=2, out_features=1, bias=True)
sigmoid : Sigmoid()
Sequential(
  (0): Linear(in_features=2, out_features=1, bias=True)
  (1): Sigmoid()
)


# 

### *0, 1을 예측하는 이진 분류 문제, Binary Cross Entropy 사용*

In [20]:
criterion = nn.BCELoss().to(device) # 비용 함수
print(criterion)

optimizer = torch.optim.SGD(model.parameters(), lr=1) # 경사 하강법 optimizer
print(optimizer)

BCELoss()
SGD (
Parameter Group 0
    dampening: 0
    lr: 1
    maximize: False
    momentum: 0
    nesterov: False
    weight_decay: 0
)


### *0 ~ 10,000 에포크 수행*

In [30]:
for step in range(10001):    
    optimizer.zero_grad() # 미분을 통해 얻은 기울기를 0으로 초기화합니다. 
    hypothesis = model(X)
    
    cost = criterion(hypothesis, Y) # 비용 함수
    cost.backward() # 비용 함수를 미분하여 gradient 계산, 가중치 W와 편향 b에 대한 기울기가 계산
    optimizer.step() # W, b를 optimizer에 업데이트
    
    if step % 100 == 0: # 100번째 epoch마다 비용 출력
        print(f"{step} step → cost: {cost.item()}")

0 step → cost: 0.6931472420692444
100 step → cost: 0.6931472420692444
200 step → cost: 0.6931472420692444
300 step → cost: 0.6931472420692444
400 step → cost: 0.6931472420692444
500 step → cost: 0.6931472420692444
600 step → cost: 0.6931472420692444
700 step → cost: 0.6931472420692444
800 step → cost: 0.6931472420692444
900 step → cost: 0.6931472420692444
1000 step → cost: 0.6931472420692444
1100 step → cost: 0.6931472420692444
1200 step → cost: 0.6931472420692444
1300 step → cost: 0.6931472420692444
1400 step → cost: 0.6931472420692444
1500 step → cost: 0.6931472420692444
1600 step → cost: 0.6931472420692444
1700 step → cost: 0.6931472420692444
1800 step → cost: 0.6931472420692444
1900 step → cost: 0.6931472420692444
2000 step → cost: 0.6931472420692444
2100 step → cost: 0.6931472420692444
2200 step → cost: 0.6931472420692444
2300 step → cost: 0.6931472420692444
2400 step → cost: 0.6931472420692444
2500 step → cost: 0.6931472420692444
2600 step → cost: 0.6931472420692444
2700 step → c

##### *loss가 0.693..로 수렴되는데 단층 퍼셉트론는 XOR 문제를 해결할 수 없다.*

# 

### *학습된 단층 퍼셉트론*

In [34]:
with torch.no_grad(): # gradient 계산 수행 x
    hypothesis = model(X)
    predicted = (hypothesis > 0.5).float()
    accuracy = (predicted == Y).float().mean()
    print('모델의 출력값(Hypothesis): ', hypothesis)
    print('모델의 출력값(Hypothesis): ', hypothesis.detach().cpu().numpy())
    print('모델의 예측값(Predicted): ', predicted.detach().cpu().numpy())
    print('실제값(Y): ', Y.cpu().numpy())
    print('정확도(Accuracy): ', accuracy.item())

모델의 출력값(Hypothesis):  tensor([[0.4998],
        [0.5000],
        [0.5000],
        [0.5002]], device='cuda:0')
모델의 출력값(Hypothesis):  [[0.49976903]
 [0.4999898 ]
 [0.5000102 ]
 [0.50023097]]
모델의 예측값(Predicted):  [[0.]
 [0.]
 [1.]
 [1.]]
실제값(Y):  [[0.]
 [1.]
 [1.]
 [0.]]
정확도(Accuracy):  0.5
